### PR TITLE
bpo-34260: fix docstring of shutil.copy2

### PR DESCRIPTION
**Pull request by csernazs**
Fix the docstring of copy2, as it does not copy file ownership (user and
group), just only mode bits, mtime, atime and flags.

The original text was confusing to developers as it suggested that this
command is the same as 'cp -p', but according to cp(1), '-p' copies file
ownership as well.

https://bugs.python.org/issue34260

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

You can [check yourself](https://devguide.python.org/committing.html#checking-if-the-cla-has-been-received)
to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Review by BoboTiG — COMMENTED** [review]
Thanks! This seems OK for me.

**Review by BoboTiG — APPROVED** [review]

**Review by pablogsal — APPROVED** [review]

**Comment by csernazs** [comment]
I've signed CLA, could you remove the _CLA not signed_ label? 
https://bugs.python.org/user?@template=clacheck&github_names=csernazs

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: I'm confused by "mode bits", I would prefer to only write "mode". I don't know what are "flags". I'm unable to find "st_flags" in os.stat_result.

I suggest to either give the full list of all copied attributes, or document which ones are not copied. But this documentation seems to be half baked.

**Review by vstinner — DISMISSED** [review]
I dislike when docstring and doc in Doc/library/shutil.rst are different:
https://docs.python.org/dev/library/shutil.html#shutil.copy2

"Identical to copy() except that copy2() also attempts to preserve all file metadata."

This documentation is wrong: the owner is not copied.

File: Lib/shutil.py
Comment (csernazs) [reply Lib/shutil.py]: I've changed the rst. I think it is written well mostly, so only the text 'all' has been replaced to 'more', which is correct now I believe.

'mode bits' has been replaced everywhere to 'mode' as you suggested. Also, I decided to write the concrete list of what metadata is copied by copystat, and copy2() refers to it to avoid unnecessary duplicating of the docstring.

st_flags is actually returned by stat() on some systems (*bsd and osx I believe), so I kept it there (I think it is not in stat_info's repr()), with the notation that flags is only copied when os.chflags() is available and st_flags is returned by stat(). More info about chflags: https://www.freebsd.org/cgi/man.cgi?query=chflags

**Comment by csernazs** [comment]
I have made the requested changes; please review again.

Please see my comments in a reply to your comment above.

File: Misc/NEWS.d/next/Library/2018-07-28-14-05-43.bpo-34260.SR5XMW.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2018-07-28-14-05-43.bpo-34260.SR5XMW.rst]: We don't document changes on the documentation. I would suggest to remove the NEWS entry.

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: I suggest to remove "some", the next sentence is enough.

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: I'm sorry, I didn't notice that the existing doc already used "mode bits". Please leave this doc unchanged ;-)

File: Doc/library/shutil.rst
Comment (vstinner) [inline Doc/library/shutil.rst]: I suggest to remove "more".

Comment (csernazs) [reply Doc/library/shutil.rst]: Problem is that shutil.copy() still copies the mode, so it is not true that the difference is that "copy2  attempts to preserve file metadata", as copy also preserves metadata.
Anyway, I removed the 'more' word as you suggested.

**Comment by csernazs** [comment]
I have made the requested changes; please review again.

Could you please also add the "skip news" label to this?
Also, the devguide would need to be updated at some point in the future as it says nothing about "no news file is needed for documentation-only changes".

**Comment by vstinner** [comment]
@giampaolo, @gpshead: would you mind to review this PR?

**Comment by giampaolo** [comment]
> I dislike when docstring and doc in Doc/library/shutil.rst are different: https://docs.python.org/dev/library/shutil.html#shutil.copy2

I agree with Victor. Doc already states it twice (once with a warning) that owner/group are not copied. 

As for this PR I find `copystat` docstring too verbose/detailed. The conciseness of the current incarnation has more value IMO. It could simply be fixed a bit:

```diff
- Copy all stat info (mode bits, atime, mtime, flags) from src to dst.
+ Copy stat info (mode bits, atime, mtime, flags, xattrs) from src to dst.
+ Owner and group are not copied.
```

As for `copy2` I suppose you can simply remove the "cp -p src dst" part since it's not true (-p preserves ownership).

**Comment by csernazs** [comment]
Alright, so I modified the copystat() docstring as you suggested.

For copy2(), I think that by only removing the "cp -p" does not solve the issue, as in this case the text would say:
```
Copy data and all stat info. Return the file's destination.
```

Which is not true as not all stat info is copied (st_uid, st_gid in this case). So I kept my changes on this.

Also I think the 'more' word would be required to the rst, saying:
```diff
-   also attempts to preserve all file metadata.
+   also attempts to preserve more file metadata.
```

**Comment by csernazs** [comment]
I have made the requested changes; please review again.

**Review by gpshead — APPROVED** [review]
looks good to me, thanks!

**Comment by csernazs** [comment]
Maybe we would also need to fix the copy() documentation as well:

> To preserve all file metadata from the original, use copy2() instead.

I'm unsure about the wording in this case.

Also I think the sentence "Identical to copy() except that copy2() also attempts to preserve file metadata." makes no sense as copy() also preserves file metadata:

> copy() copies the file data and the file’s permission mode (see os.chmod()).

 so both copy() and copy2() preserve file metadata, copy() preserves only mode, copy2() preserves mode, atime, mtime, etc..

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: I prefer the phrasing from shutils.rst: "The file contents, owner, and group are unaffected." (shorter and obvious)

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: xattrs: the shutil.rst doc gives more detail: "On Linux, copystat() also copies the “extended attributes” where possible." I understand that xattrs are not always copied, thing which isn't as well explained here.

I suggest to use "Copy file metadata" for the docstring header, and then copy shutil.rst doc which is good:

"Copy the permission bits, last access time, last modification time, and flags from src to dst. On Linux, copystat() also copies the “extended attributes” where possible. The file contents, owner, and group are unaffected. src and dst are path names given as strings."

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by csernazs** [comment]
I have made the requested changes; please review again

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
I close/reopen the issue to re-trigger CI jobs. GitHub is sick today, https://status.github.com/messages

**Comment by vstinner** [comment]
"AppVeyor build failed (...) Console: Log is empty." thank you AppVeyor...

**Comment by vstinner** [comment]
I rescheduled the AppVeyor job. If AppVeyor doesn't pass, I cannot merge the PR.

**Comment by vstinner** [comment]
@csernazs: Oh, the backport failed. Can you please try to backport the change using " cherry_picker 4f399be0e70d8b5516b6213568b7665765bb3114 3.7" (see previous comment) or using "git cherry-pick -x 4f399be0e70d8b5516b6213568b7665765bb3114"?

IMHO this change also deserves to be backported to 2.7 and 3.6 as well.

**Pull request merged**
